File: lib/utils/sun_calculator.dart
import 'package:sunrise_sunset_calc/sunrise_sunset_calc.dart';

/// Utilidad para calcular eventos solares (amanecer, atardecer) basados en ubicación y fecha.
class SunCalculator {
  /// Calcula los tiempos solares para una fecha y ubicación dadas usando un paquete probado.
  static SunTimes calculateTimes(DateTime date, double lat, double lng) {
    // Calculamos el amanecer y atardecer en UTC puro para evitar doble suma horaria
    final result = getSunriseSunset(lat, lng, const Duration(seconds: 0), date.toUtc());
    
    // Luego convertimos el resultado UTC puro a la zona horaria del dispositivo
    return SunTimes(
      sunrise: result.sunrise.toLocal(),
      sunset: result.sunset.toLocal(),
    );
  }
}

class SunTimes {
  final DateTime sunrise;
  final DateTime sunset;

  SunTimes({required this.sunrise, required this.sunset});
  
  @override
  String toString() => 'Sunrise: $sunrise, Sunset: $sunset';
}

File: app/src/main/java/com/nubo/nubo/domain/astro/SunCalculator.kt
package com.nubo.nubo.domain.astro

import java.time.Instant
import java.time.LocalDateTime
import java.time.ZoneId

/** Amanecer y ocaso de un día concreto, en la zona horaria del dispositivo. */
data class SunTimes(
    val sunrise: LocalDateTime,
    val sunset: LocalDateTime,
)

/** Calcula los eventos solares de una fecha y unas coordenadas. */
object SunCalculator {

    /**
     * Devuelve el amanecer y el ocaso del día de [dateTime] en [lat]/[lng].
     *
     * En latitudes extremas puede no haber salida ni puesta (día o noche
     * polar); ahí se devuelve `null` y quien llame debe decidir qué mostrar.
     */
    fun calculateTimes(
        dateTime: LocalDateTime,
        lat: Double,
        lng: Double,
        zone: ZoneId = ZoneId.systemDefault(),
    ): SunTimes? {
        // Se calcula sobre el mediodía local para que el día natural quede
        // centrado en la ventana y no se cuele el amanecer del día contiguo.
        val noon: Instant = dateTime.toLocalDate()
            .atTime(12, 0)
            .atZone(zone)
            .toInstant()

        val (sunrise, sunset) = SunCalc.sunTimes(noon, lat, lng)
        if (sunrise == null || sunset == null) return null

        return SunTimes(
            sunrise = LocalDateTime.ofInstant(sunrise, zone),
            sunset = LocalDateTime.ofInstant(sunset, zone),
        )
    }
}
